rlog/2024-12-08-merkle-tree.mdx

```diff
@@ -0,0 +1,197 @@
+A hash function is collision resistant; it is difficult for a malicious party to find two values share the same hash value.
+This collision resistance property, essentially, fixes the input to each leaf and into their parent, their parent's parent, and so on.
+
+In certain applications,
+it may be desirable for the digest of a Merkle tree to be kept confidential.
+This is achieved with the preimage resistant propery of hash functions.
+A hash function is preimage resistant provided that it is difficult reverse the hashing operation. 
+It would be necessary for a malicious party to find preimages to each node starting from the root node to determine original digest.
+
+Now, we see that Merkle trees are secured structures that are tamper resistant.
+
+### Proof of membership
+An interesting and critical use of Merkle trees is its ability to prove any piece of data is a part of its digest.
+This can be done with logarthmic storage and logartihmic computation time.
+
+Suppose that we want to show that data $\ell$ is part of the Merkle tree's digest.
+Additionally, suppose that $\mathsf{hash}$ is the hash function used to construct the tree.
+We assume that the hash function $\mathsf{hash}$ can be computed in constant-time for any input.
+
+Suppose that the Prover provides data $\ell$ to the Verifier,
+and tells the Verifier that $\ell$ corresponds to the $i$th leaf of the Merkle tree.
+For the Verifier to be convinced that $\ell$ is part of the digest, he needs to be able to construct the tree's root hash using $\mathsf{hash}$, $i$, $\ell$ and some additional information from the Prover.
+Specifically, the Prover must provide the sibling hashes for each value that the Verifier can compute.
+This enables the Verifier to compute the parents of the siblings that the Prover provides and the values that he was able to produce himself.
+
+The leaf index $i$ indicates whether a hash value provided by the Prover is a left or right sibling.
+This is done by looking at the binary expansion of $i$.
+
+The Verifier can compute the leaf $h_0 = \mathsf{hash}(\ell)$.
+Next, using $h_0$'s sibling, $h'_0$, provided by the Prover:
+the Verifier can compute $h_1 = \mathsf{hash}(h_0 \|h'_0)$ or $h_1 = \mathsf{hash}(h'_0 \| h_0)$
+depending on whether $h'_0$ is a left or right sibling.
+This pathing continues until the Verifier either successfully computes the root hash (in $n+1$ hashes) or fails to.
+
+The Prover has to provide $n$ sibling nodes for the proof of membership.
+
+There is a key detail that is essential for the proof of membership to be secure.
+The root hash has to be provided to the Verifier prior to the selection of data $\ell$.
+Otherwise, Prover could generate a series of hash values (with the corresponding root hash) to forge a proof of membership.
+
+## Extensions of Merkle trees
+
+Merkle trees can be extended in multiple ways.
+In this section, we explore a select few of these extensions.
+
+### Sparse Merkle trees
```

Comment:
adding some white-listed implementations can be good

Reply:
Added a repo for Verkle tree. Cannot locate a white-listed implementation of sparse merkle trees.